design-proposals/klip-19-materialize-views.md

```diff
@@ -0,0 +1,223 @@
+# KLIP 19 - Introduce Materialize Views
+
+**Author**: @big-andy-coates | 
+**Release Target**: TBD | 
+**Status**: In Discussion | 
+**Discussion**: TBD
+
+**tl;dr:** 
+KSQL currently allows users to insert into any Kafka topic. In the case of table data this can be semantically incorrect,
+leading to bugs and poor UX. This limitation can force organisations to rely on Kafka ACLs to limit such actions. 
+However, ACLs should be used for access control, not to enforce correct semantics. Introducing Materialized Views will 
+clean up the semantics around `INSERT VALUES` into tables and allow users to express intended mutability of data sources.    
+
+## Motivation and background
+
+### Why might it be bad to allow inserts into tables?
+
+Let's consider two examples:
+1. A table in an upstream db is being replicated to a topic in Kafka, e.g. via Connect, and users want this 
+   table available in KSQL.
+1. A user wants to create a table in KSQL that is derived from other source(s), as currently done using a 
+   `CREATE TABLE AS SELECT` statement.
+   
+In the first example, the _source of truth_ for the data set is the upstream db, not in Kafka. Any change to the data 
+needs to be done in the upstream db. Simply writing a new row to the Kafka topic is semantically incorrect. The row will
+be overwritten if/when the same row in the database is next updated. What's more, that same topic may be used by other 
+systems that expect it to be an exact copy of the data in the database table. 
+
+In the second example, the rows written to the sink topic are a _derived dataset_.  Simply writing a new row to the Kafka 
+topic is also semantically incorrect. Any row written would potentially be overwritten by an upstream change.
+
+### When is it OK to insert into a table?
+
+We _can_ allow inserts into tables where KSQL is the only writer of that data and we can ensure correct table semantics.
+
+At the moment, users can create a new table in KSQL using a `CREATE TABLE` statement with `PARTITIONS` and, optionally,
+`REPLICAS`. This will create a new topic. As no other system writes to this topic, KSQL can support `INSERT`s into it.
+Likewise, a `CREATE TABLE` statement can be used to import an existing topic into KSQL. As long as no other system is 
+writing to that topic, (and that includes another KSQL topology), KSQL can offer correct `INSERT` semantics.
+
+The defining factor here is that it is only possible to have correct `INSERT` semantics _if no other system updates the table_.
+
+**Side note**: It would be easy to support correct `DELETE` semantics too. However, correct `UPDATE` row semantics will
+require more work. Likely, it will involve proxying calls to the KSQL node that owns the key, ensuring a single writer
+pattern.
```

Comment:
Why is that required?

Reply:
To implement `UPDATE` we need to do an atomic read-update-write of the value. Likely, we'd implement this by materializing the table into rocksdb, which would mean a specific row is owned by a specific KSQL node. Hence any `UPDATE` to that row would need to be routed to that node.

Reply:
Well, but don't we have the same issue with `INSERT`? If inserts are async as described, two insert for the same primary key would happen in parallel, but only the first one would be "valid". (`DELETE` would not have this issue as it would be idempotent.)

Reply:
The `INSERT` is more correctly an `UPSERT` (We should probably correct that syntax too).  Each `INSERT` fully replaces any previous value.

However, an `UPDATE` may only provide partial information i.e. `UPDATE foo SET col1 = 1 WHERE col2 = 3;`. This requires atomic read-update-writes.

I've added more details to the KLIP.